src/Runner/Runner.php

```diff
@@ -89,26 +119,245 @@ public function __construct(
+            $decoder->on('data', static function (array $data) use ($processPool, $fileChunk, $decoder, $encoder): void {
+                if (ParallelAction::RUNNER_HELLO !== $data['action']) {
+                    return;
+                }
+
+                $identifier = ProcessIdentifier::fromRaw($data['identifier']);
+                $process = $processPool->getProcess($identifier);
+                $process->bindConnection($decoder, $encoder);
+                $job = $fileChunk();
+
+                if (0 === \count($job)) {
+                    $process->request(['action' => ParallelAction::WORKER_THANK_YOU]);
+                    $processPool->endProcessIfKnown($identifier);
+
+                    return;
+                }
+
+                $process->request(['action' => ParallelAction::WORKER_RUN, 'files' => $job]);
+            });
+        });
+
+        $processesToSpawn = min(
+            $this->parallelConfig->getMaxProcesses(),
+            (int) ceil($this->fileCount / $this->parallelConfig->getFilesPerProcess())
         );
+        $processFactory = new ProcessFactory($this->input);
+
+        for ($i = 0; $i < $processesToSpawn; ++$i) {
+            $identifier = ProcessIdentifier::create();
+            $process = $processFactory->create(
+                $streamSelectLoop,
+                new RunnerConfig(
+                    $this->isDryRun,
+                    $this->stopOnViolation,
+                    $this->parallelConfig,
+                    $this->configFile
+                ),
+                $identifier,
+                $serverPort,
+            );
+            $processPool->addProcess($identifier, $process);
+            $process->start(
+                // [REACT] Handle workers' responses (multiple actions possible)
+                function (array $workerResponse) use ($processPool, $process, $identifier, $fileChunk, &$changed): void {
+                    // File analysis result (we want close-to-realtime progress with frequent cache savings)
+                    if (ParallelAction::RUNNER_RESULT === $workerResponse['action']) {
+                        $fileAbsolutePath = $workerResponse['file'];
+                        $fileRelativePath = $this->directory->getRelativePathTo($fileAbsolutePath);
+
+                        // Pass-back information about applied changes (only if there are any)
+                        if (isset($workerResponse['fixInfo'])) {
+                            $changed[$fileRelativePath] = $workerResponse['fixInfo'];
+                        }
+                        // Dispatch an event for each file processed and dispatch its status (required for progress output)
+                        $this->dispatchEvent(FixerFileProcessedEvent::NAME, new FixerFileProcessedEvent($workerResponse['status']));
+
+                        if (
+                            FixerFileProcessedEvent::STATUS_NO_CHANGES === (int) $workerResponse['status']
```

Comment:
can worker response have proper types in first place ? so we do not need to cast types again and again

Reply:
How do you want to achieve this? This is decoded JSON that comes from the socket.